contracts/access/AccessControl.sol

```diff
@@ -0,0 +1,177 @@
+pragma solidity ^0.6.0;
+
+import "../utils/EnumerableSet.sol";
+
+/**
+ * @dev Contract module that allows children to implement role-based access
+ * control mechanisms.
+ *
+ * Roles are referred to by their `bytes32` identifier. These should be exposed
+ * in the external API and be unique. The best way to achieve this is by
+ * using `public constant` hash digests:
+ *
+ * ```
+ * bytes32 public constant MY_ROLE = keccak256("MY_ROLE");
+ * ```
+ *
+ * Roles can be used to represent a set of permissions. To restrict access to a
+ * function call, use {hasRole}:
+ *
+ * ```
+ * function foo() public {
+ *     require(hasRole(MY_ROLE, msg.sender));
+ *     ...
+ * }
+ * ```
+ *
+ * Roles can be granted and revoked programatically by calling the `internal`
+ * {_grantRole} and {_revokeRole} functions.
+ *
+ * This can also be achieved dynamically via the `external` {grantRole} and
+ * {revokeRole} functions. Each role has an associated admin role, and only
+ * accounts that have a role's admin role can call {grantRole} and {revokeRoke}.
+ *
+ * By default, the admin role for all roles is `DEFAULT_ADMIN_ROLE`, which means
+ * that only accounts with this role will be able to grant or revoke other
+ * roles. More complex role relationships can be created by using
+ * {_setRoleAdmin}.
+ */
+abstract contract AccessControl {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    struct Role {
+        EnumerableSet.AddressSet members;
+        bytes32 admin; // This role's admin role
+    }
+
+    mapping (bytes32 => Role) private _roles;
+
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    /**
+     * @dev Emitted when `account` is granted `roleId` role.
+     */
+    event RoleGranted(bytes32 indexed roleId, address indexed account);
```

Comment:
@frangio we've discussed what name the `bytes32` entity should have (is it role? role id?). In that vein, what do you think about the `roleId` event and function argument?

Reply:
I prefer `role` because I'm allergic to camel case... But this is not a great reason. I don't have strong reasons either way. I think the fact that we are saying "a role is the struct" shouldn't matter for this because the struct is only used internally. Is there any distinction we would want to make between the "role id" and the "role" (whatever that is? is it even anything?).

Reply:
Yeah, I think I'll change just everything to `role` instead and try to do away with the 'id' naming.